internal/service/alarm_notification_handler.go

```diff
@@ -0,0 +1,287 @@
+	b.logger = value
+	return b
+}
+
+// SetLoggingWrapper sets the wrapper that will be used to configure logging for the HTTP clients
+// used to connect to other servers, including the backend server. This is optional.
+func (b *alarmNotificationHandlerBuilder) SetLoggingWrapper(
+	value func(http.RoundTripper) http.RoundTripper) *alarmNotificationHandlerBuilder {
+	b.loggingWrapper = value
+	return b
+}
+
+// SetCloudID sets the identifier of the O-Cloud of this handler. This is mandatory.
+func (b *alarmNotificationHandlerBuilder) SetCloudID(
+	value string) *alarmNotificationHandlerBuilder {
+	b.cloudID = value
+	return b
+}
+
+// SetExtensions sets the fields that will be added to the extensions.
+func (b *alarmNotificationHandlerBuilder) SetExtensions(
+	values ...string) *alarmNotificationHandlerBuilder {
+	b.extensions = values
+	return b
+}
+
+// SetExtensions sets the fields that will be added to the extensions.
+func (b *alarmNotificationHandlerBuilder) SetKubeClient(
+	kubeClient *k8s.Client) *alarmNotificationHandlerBuilder {
+	b.kubeClient = kubeClient
+	return b
+}
+
+// Build uses the data stored in the builder to create anad configure a new handler.
+func (b *alarmNotificationHandlerBuilder) Build(ctx context.Context) (
+	result *alarmNotificationHandler, err error) {
+	// Check parameters:
+	if b.logger == nil {
+		err = errors.New("logger is mandatory")
+		return
+	}
+	if b.cloudID == "" {
+		err = errors.New("cloud identifier is mandatory")
+		return
+	}
+
+	if b.kubeClient == nil {
+		err = errors.New("kubeClient is mandatory")
+		return
+	}
+
+	// Prepare the JSON iterator API:
+	jsonConfig := jsoniter.Config{
+		IndentionStep: 2,
+	}
+	jsonAPI := jsonConfig.Froze()
+
+	// Create the filter expression evaluator:
+	pathEvaluator, err := search.NewPathEvaluator().
+		SetLogger(b.logger).
+		Build()
+	if err != nil {
+		return
+	}
+	selectorEvaluator, err := search.NewSelectorEvaluator().
+		SetLogger(b.logger).
+		SetPathEvaluator(pathEvaluator.Evaluate).
+		Build()
+	if err != nil {
+		return
+	}
+	// Create the jq tool:
+	jqTool, err := jq.NewTool().
+		SetLogger(b.logger).
+		Build()
+	if err != nil {
+		return
+	}
+
+	// Check that extensions are at least syntactically valid:
+	for _, extension := range b.extensions {
+		_, err = jqTool.Compile(extension)
+		if err != nil {
+			return
+		}
+	}
+
+	alarmSubscriptionSearcher := newAlarmSubscriptionSearcher()
+	alarmSubscriptionSearcher.SetLogger(b.logger).SetJqTool(jqTool).build()
+
+	// create persist storeage option
+	persistStore := persiststorage.NewKubeConfigMapStore().
+		SetNameSpace(TestNamespace).
+		SetName(TestConfigmapName).
+		SetFieldOwnder(FieldOwner).
+		SetJsonAPI(&jsonAPI).
+		SetClient(b.kubeClient)
+
+	// http client to send out notification
+	// use 2 sec first
+	httpClient := http.Client{Timeout: 2 * time.Second}
+
+	if singleAlarmNotificationHandle != nil {
+		panic("There is an existing handler instance")
+	}
```

Comment:
I think we should not do this check: it is perfectly OK to create multiple alarm notification handlers. Maybe one will receive notifications from Prometheus alert manager and another will receive them from something else, a hardware manager, for example.

If we really needed to ensure that there is only one instance it shouldn't be at this level, but at the level of the code that creates the handler. And to do it correctly we would need to be careful with how we do the check: using a nil check like here doesn't work well when there are multiple goroutines working in parallel.

But lets forget about that: just remove the check, I'd say.

Reply:
Ah, and we shouldn't be using "panic" anyhow, better return an error.

Reply:
I made it singleton for the configmap update call back. It is also workable to make global array/arrays. Not big deal.

For now I will change not panic but log an error. (change to handle multiple instances only when needed)

Reply:
Changed